Geb 8

matrei opened the pull request:
Closes #15065

matrei committed:
fix(deps): `geb` v. `7.0` -> `8.0.0`

matrei committed:
Merge branch '7.0.x' into geb-8

matrei committed:
fix: adapt to changes in Geb 8

Adapt to changed driver creation logic in Geb 8.

matrei committed:
refactor(geb): simplifications

matrei committed:
docs(geb): improve comments and javadoc

matrei committed:
chore(geb): whitespace

matrei commented:
@JonasPammer Would you like to look over these changes?

jdaugherty commented on build.gradle:
```diff
@@ -70,6 +70,7 @@ allprojects {
             url = 'https://repository.apache.org/content/groups/staging'
             content {
                 includeModuleByRegex('org[.]apache[.]grails[.]gradle', 'grails-publish')
+                includeModuleByRegex('org[.]apache[.]groovy[.]geb', 'geb-.*')
```
I thought you could only have one of these includes?  does this effectively or the expressions?

matrei replied:
Yes, I believe this should create two rules with logical OR.

jdaugherty replied:
Wow, so it does:  https://docs.gradle.org/current/javadoc/org/gradle/api/artifacts/repositories/RepositoryContentDescriptor.html 

I had no idea!

jdaugherty commented on grails-geb/src/testFixtures/groovy/grails/plugin/geb/WebDriverContainerHolder.groovy:
```diff
@@ -162,18 +162,10 @@ class WebDriverContainerHolder {
+        // the `RemoteWebDriver` will be instantiated using the `webdriver.remote.server` system property,
+        // so we need to set that value.
         System.setProperty('webdriver.remote.server', currentContainer.seleniumAddress.toString())
-        gebConfig.driver // This will implicitly call `createDriver()`
```
Wasn't the intent of this code to force the test container always?

matrei replied:
> Wasn't the intent of this code to force the test container always?

I think the intent was to force using the `currentContainer.seleniumAddress` when creating the driver.

JonasPammer replied:
As the cautious/unecessary restore was removed, it will just happen at any the next call to [.driver](https://github.com/apache/groovy-geb/blob/4266eeeea6f59e4718339b93283327bf176563b9/module/geb-core/src/main/groovy/geb/Browser.groovy#L172)

matrei committed:
fix(geb): reset `webdriver.remote.server` after `RemoteWebDriver` init

Reinstate resetting the system property after driver construction.

matrei committed:
docs(geb): minor comment improvement

matrei committed:
fix(geb): prevent cross-thread leakage of `webdriver.remote.server`

Scope the `webdriver.remote.server` System property to the current thread
during `RemoteWebDriver` creation, then restore it. This avoids routing
later sessions to the wrong endpoint in parallel builds.

JonasPammer commented:
now this is what i call generics, haha. I would not even know half of these functions and groovy/thread-nanigans, especially after having used solely django at work for last few months. :shushing_face:  :stuck_out_tongue_winking_eye: 

as long as the system env is set for when https://github.com/apache/groovy-geb/blob/4266eeeea6f59e4718339b93283327bf176563b9/module/geb-core/src/main/groovy/geb/Browser.groovy#L172 is called its still the same thing as my original thought/impl. :+1:

matrei commented:
> now this is what i call generics, haha. I would not even know half of these functions and groovy/thread-nanigans, especially after having used solely django at work for last few months. 🤫 😜
> 
> as long as the system env is set for when https://github.com/apache/groovy-geb/blob/4266eeeea6f59e4718339b93283327bf176563b9/module/geb-core/src/main/groovy/geb/Browser.groovy#L172 is called its still the same thing as my original thought/impl. 👍

Haha, I thought it might be overkill as well, but what if you mix `ContainerGebSpec` and regular `GebSpec` tests in the same project?

However, I don't think our GebConfig integration is working perfectly yet. I tried switching to `new RemoteWebDriver(new FirefoxOptions())` and it blows up as the container is browser specific and seems to default to chrome.
```console
Could not start a new session. Response code 500. Message: Could not start a new session. No nodes support the capabilities in the request
Host info: host: 'f1b81917b1f3', ip: '172.17.0.4'
Build info: version: '4.34.0', revision: '2a4c61c498'
System info: [os.name](http://os.name/): 'Linux', os.arch: 'amd64', os.version: '6.10.14-linuxkit', java.version: '21.0.8'
Driver info: driver.version: unknown
Build info: version: '4.34.0', revision: '707dcb4246*'
System info: [os.name](http://os.name/): 'Linux', os.arch: 'amd64', os.version: '6.8.0-79-generic', java.version: '17.0.14'
Driver info: org.openqa.selenium.remote.RemoteWebDriver
Command: [null, newSession {capabilities=[Capabilities {acceptInsecureCerts: true, browserName: firefox, grails:gebConfigUsed: true, moz:firefoxOptions: {prefs: {remote.active-protocols: 1}}}]}]
Capabilities {acceptInsecureCerts: true, browserName: firefox, grails:gebConfigUsed: true, moz:firefoxOptions: {prefs: {remote.active-protocols: 1}}}
org.openqa.selenium.SessionNotCreatedException: Could not start a new session. Response code 500. Message: Could not start a new session. No nodes support the capabilities in the request
```

This is now fixed in [fcc4d82](https://github.com/apache/grails-core/pull/15067/commits/fcc4d821af41f36b53a4f20428f7fa6693ea032b)

matrei committed:
fix(geb): allow testing with different browsers

This allows using different browsers in `GebConfig.groovy`
A `configuredBrowser` property has to be set that matches
the browser capabilities used to create the `RemoteWebDriver`
to start the correct container for the browser.

matrei committed:
chore(geb): formatting

matrei committed:
chore(geb): formatting

matrei committed:
refactor(geb): simplify system property override

matrei committed:
docs(geb): improve javadoc and code comments

matrei committed:
fix(geb): resolve trait method conflicts via explicit override

matrei committed:
refactor(geb): simplify

matrei committed:
chore(geb): formatting

matrei committed:
refactor(geb): simplify

matrei committed:
test(geb): update test label

matrei committed:
docs(geb): add `GebConfig` instructions to README

JonasPammer commented on grails-geb/src/testFixtures/groovy/grails/plugin/geb/WebDriverContainerHolder.groovy:
```diff
@@ -245,19 +268,47 @@ class WebDriverContainerHolder {
+        DockerImageName.parse("selenium/standalone-$configuredBrowser:$seleniumVersion")
+    }
+
+    private static void validateConfiguredBrowser(Object browser) {
```
why only allow a hardcoded subset?

matrei replied:
At the moment, these three seem to be the only ones that have supporting container images: https://java.testcontainers.org/modules/webdriver_containers/#other-browsers

JonasPammer replied:
Yeah but don't know if we should impose a restriction here, for it could change or disallow adaptability by another.

matrei replied:
Good point! Pushed changes.

JonasPammer commented on grails-geb/src/testFixtures/groovy/grails/plugin/geb/WebDriverContainerHolder.groovy:
```diff
@@ -59,99 +64,130 @@ import grails.plugin.geb.serviceloader.ServiceRegistry
+            // Prepare for creating a suitable container matching the driver
+            // configured in GebConfig.groovy, or more specifically,
+            // specified by the `configuredBrowser` property.
+            dockerImageName = createDockerImageName(gebConf.rawConfig.configuredBrowser)
```
regression for GebConfig's that don't have this special var - add a default here?
we already imply Chrome at `new RemoteWebDriver(container.seleniumAddress, new ChromeOptions()` when no GebConfig existed on classpath

matrei replied:
Missing `configuredBrowser` property in `GebConfig.groovy` will result in error message:
```console
Illegal 'configuredBrowser' property value in GebConfig.groovy: [null]. Valid values are [chrome, edge, firefox]
```

I don't think we can have a default for `configuredBrowser`, as that will risk the container not matching the configured driver.
Are you suggesting we revert/default to chrome driver and container if the `GebConfig.groovy` is missing the `configuredBrowser` property?

JonasPammer replied:
Yes, default to chrome.
If possible, put a try-catch where the "could not find container matching driver restriction" appears, and put this message as a possible suggestion/fix from us before the printStackTrace.

matrei replied:
Pushed some changes, have a look.

JonasPammer replied:
I didn't mean for us to force an own RemoteWebDriver with ChromeOptions **just** because containerBrowser property is missing, but to (pseudo code - need to find the spot)

```groovy
                } catch (SessionNotCreatedException e){
                    log.error("Failed to create remote browser session: $e.message. Did you set a 'containerBrowser' property corresponding to the driver in GebConfig.groovy?", e)
                }
```

which probably will imply removing the gebConfigFileActive logic

JonasPammer replied:
[nice](https://github.com/apache/grails-core/pull/15067/commits/79b8708c8f39b4f77d11ce7c124dd9ac4ee246e6)!

JonasPammer commented on grails-geb/README.md:
```diff
@@ -161,6 +161,28 @@ To enable tracing, set the following system property:
+
+To make this work, ensure:
+1. The `driver` property in your `GebConfig` is a `Closure` that returns a `RemoteWebDriver` instance.
+2. You set a custom `configuredBrowser` property so that `ContainerGebSpec` can start a matching container (e.g. "chrome", "firefox", "edge").
```
as pointed out some projects potentially have GebConfig next to ContainerGebConfig. I saw other extensions always prefix their variables with their extension (e.g. sauceLabsBrowser), maybe we should too to avoid confusion? e.g. `containerBrowser` (in reference to containergebspec)

matrei replied:
Yes, that is a better name!

JonasPammer reviewed (commented):
even opening grails-core brtings down my pc at home, much more when running tests, so could only do github overview

matrei committed:
fix(geb): make inner classes `static`

matrei committed:
feedback(geb): rename property to `containerBrowser`

matrei committed:
chore(geb): formatting

matrei committed:
fix(geb): add message for missing `containerBrowser`

Add a more helpful error message in the case where
a `containerBrowser` property is missing from `GebConfig.groovy`.

matrei committed:
test(geb): make tests more resilient in slow runners

JonasPammer commented:
> However, I don't think our GebConfig integration is working perfectly yet. I tried switching to `new RemoteWebDriver(new FirefoxOptions())` and it blows up as the container is browser specific and seems to default to chrome.
> 
> ```
> Could not start a new session. Response code 500. Message: Could not start a new session. No nodes support the capabilities in the request
> Host info: host: 'f1b81917b1f3', ip: '172.17.0.4'
> Build info: version: '4.34.0', revision: '2a4c61c498'
> System info: [os.name](http://os.name/): 'Linux', os.arch: 'amd64', os.version: '6.10.14-linuxkit', java.version: '21.0.8'
> Driver info: driver.version: unknown
> Build info: version: '4.34.0', revision: '707dcb4246*'
> System info: [os.name](http://os.name/): 'Linux', os.arch: 'amd64', os.version: '6.8.0-79-generic', java.version: '17.0.14'
> Driver info: org.openqa.selenium.remote.RemoteWebDriver
> Command: [null, newSession {capabilities=[Capabilities {acceptInsecureCerts: true, browserName: firefox, grails:gebConfigUsed: true, moz:firefoxOptions: {prefs: {remote.active-protocols: 1}}}]}]
> Capabilities {acceptInsecureCerts: true, browserName: firefox, grails:gebConfigUsed: true, moz:firefoxOptions: {prefs: {remote.active-protocols: 1}}}
> org.openqa.selenium.SessionNotCreatedException: Could not start a new session. Response code 500. Message: Could not start a new session. No nodes support the capabilities in the request
> ```
> 
> This is now fixed in [fcc4d82](https://github.com/apache/grails-core/pull/15067/commits/fcc4d821af41f36b53a4f20428f7fa6693ea032b)

Weird, because [it](https://github.com/testcontainers/testcontainers-java/blob/51219646dca72ad267e575bf25d0b60208c60b42/modules/selenium/src/main/java/org/testcontainers/containers/BrowserWebDriverContainer.java#L195) uses [getStandardImageForCapabilities](https://github.com/testcontainers/testcontainers-java/blob/51219646dca72ad267e575bf25d0b60208c60b42/modules/selenium/src/main/java/org/testcontainers/containers/BrowserWebDriverContainer.java#L246)
Will try to reproduce

----

Update, i forgot about the chicken and egg problem. Can't call `.withCapabilities(((RemoteWebDriver)gebConf.createDriver()).getCapabilities())` because it tries to connect (`java.lang.NullPointerException: Cannot invoke "org.testcontainers.containers.BrowserWebDriverContainer.getSeleniumAddress()" because "this.container" is null`). Hiding

matrei committed:
fix(feedback): skip `containerBrowser` validation

matrei committed:
refactor(geb): extract methods

matrei committed:
fix(geb): move `atCheckWaiting` to CI env variable

JonasPammer commented on grails-geb/src/testFixtures/groovy/grails/plugin/geb/WebDriverContainerHolder.groovy:
```diff
@@ -243,6 +244,9 @@ class WebDriverContainerHolder {
             browser.driver.manage().timeouts().pageLoadTimeout(Duration.ofSeconds(settings.pageLoadTimeout))
         if (settings.scriptTimeout != DEFAULT_TIMEOUT_SCRIPT)
             browser.driver.manage().timeouts().scriptTimeout(Duration.ofSeconds(settings.scriptTimeout))
+        if (settings.atCheckWaiting != DEFAULT_AT_CHECK_WAITING) {
```
to be honest, now that we read GebConfig like normal Browser instantiaton, I don't even  know if we should keep our GebConfig'able browser/driver settings as part of our custom GrailsGebSettings (this includes for example timeouts).

i.e. rather than cc17f14a48eebd897116408c371a94300e943c94
```
        env:
          # Make Geb tests more resilient in slow CI environments
          JAVA_TOOL_OPTIONS: -Dgrails.geb.atCheckWaiting=true
```

do

`grails-test-examples/geb-gebconfig/src/integration-test/resources/GebConfig.groovy`
```groovy
atCheckWaiting = System.getenv('CI') != null
```

matrei replied:
The good thing with `GrailsGebSettings` is that we can use system properties without changes or additions to source code.
I wish this was possible with stock Geb.

The option to use `GebConfig` is also available.

JonasPammer replied:
ok, i'd think if the user wants that he could also add system property code in his gebconfig but sure does not hurt, only will be weird when some things are settable and some not through this dep.

matrei replied:
My thinking is, that it should not be necessary to have a `GebConfig` to run `ContainerGebSpec`s (convention over configuration).
For the `grails-core` functional tests for example, we have no `GebConfigs` (except one).
And if we wanted to set `atCheckWaiting=true` we would have to add a `GebConfig` to each functional test project.
Now we can just set a system property in CI.

> only will be weird when some things are settable and some not through this dep

I agree, but this is already the case with stock Geb: https://groovy.apache.org/geb/manual/current/#system-properties

JonasPammer commented on grails-geb/src/testFixtures/groovy/grails/plugin/geb/GrailsGebSettings.groovy:
```diff
@@ -66,15 +68,21 @@ class GrailsGebSettings {
         implicitlyWait = getIntProperty('grails.geb.timeouts.implicitlyWait', DEFAULT_TIMEOUT_IMPLICITLY_WAIT)
         pageLoadTimeout = getIntProperty('grails.geb.timeouts.pageLoad', DEFAULT_TIMEOUT_PAGE_LOAD)
         scriptTimeout = getIntProperty('grails.geb.timeouts.script', DEFAULT_TIMEOUT_SCRIPT)
+        atCheckWaiting = getBooleanProperty('grails.geb.atCheckWaiting', DEFAULT_AT_CHECK_WAITING)
```
even if we wanna keep this GrailsGebSetting (other comment), it accepts more than a boolean - https://groovy.apache.org/geb/manual/current/#at-check-waiting

matrei replied:
Yes, but I thought it was overkill to implement logic for handling that.

JonasPammer replied:
Shouldn't it just be getStringProperty? no need for own validation of correctness, it will fail later if incorrect anyway too

matrei replied:
Wouldn't a String imply a "wait preset name"? https://groovy.apache.org/geb/manual/current/#content-dsl-wait

JonasPammer replied:
ah ok yes i thought it would just pick the right one with trial and error. yes your implementation is best as we set the default timeouts with environment too! :+1:

matrei committed:
fix(geb): feedback - remove container fallback

Throw exception when `driver` and `containerBrowser`
are mismatched in `GebConfig` instead of falling
back to chrome.

matrei committed:
fix(geb): system property setting of timeout values

matrei committed:
ci: update Geb `atCheckWaiting` system property

matrei committed:
chore(geb): formatting

JonasPammer reviewed (approved):
Great Thanks

jdaugherty commented on .github/workflows/gradle.yml:
```diff
@@ -173,6 +173,9 @@ jobs:
       - name: "🏃 Run Functional Tests"
+        env:
+          # Make Geb tests more resilient in slow CI environments
+          JAVA_TOOL_OPTIONS: -Dgrails.geb.atCheckWaiting.enabled=true
```
Why make CI work differently then locally?  Does our JDK read this by default?  I would have assumed we would have defined this property as part of the test functional test configuration instead.

jdaugherty replied:
I think this is my only concern. This seems better suited in the gradle files than in the CI configuration.

matrei replied:
> Why make CI work differently then locally?

It should only be a concern in low resource environments, to not get false positives because the runner can't serve the page fast enough. We see these random failures in CI.

> Does our JDK read this by default?4

Gradle will automatically pass through the `JAVA_TOOL_OPTIONS` to the JVM, without needing `systemProperty 'grails.geb.atCheckWaiting.enabled', System.getProperty('grails.geb.atCheckWaiting.enabled')` in the build file.

>  I would have assumed we would have defined this property as part of the test functional test configuration instead.

We could do that, I think this way is more flexible.

jdaugherty replied:
Can we at least add a comment to the functional test config so we know this configuration is present / different in CI?

matrei replied:
I moved setting `atCheckWaiting` to `functional-test-config.gradle`.

jdaugherty commented on grails-geb/README.md:
```diff
@@ -161,6 +177,28 @@ To enable tracing, set the following system property:
 This allows you to opt in to tracing when an OpenTelemetry collector is available.
 
+#### GebConfig.groovy and using non-default browser settings
+Provide a `GebConfig.groovy` on the test runtime classpath (commonly `src/integration-test/resources`, but any location on the test classpath works) to customize the browser.
```
I assume this GebConfig is optional?

matrei replied:
> I assume this GebConfig is optional?

Yes, `GebConfig` is optional.

jdaugherty commented on grails-geb/src/testFixtures/groovy/grails/plugin/geb/GrailsGebSettings.groovy:
```diff
@@ -65,17 +73,48 @@ class GrailsGebSettings {
+        if (propValue) {
+            try {
+                if (propValue.contains('.')) {
+                    return Double.parseDouble(propValue)
```
Does the geb api really use Doubles and not BigDecimals?

matrei replied:
> Does the geb api really use Doubles and not BigDecimals?

The Geb API uses `Number`. Should I create a `BigDecimal` instead of a `double`?

jdaugherty replied:
I would have assumed BigDecimal is a better default

matrei replied:
Changed to `BigDecimal`.

matrei committed:
fix(geb): use correct `atCheckWaiting` system property

jdaugherty commented on grails-geb/src/testFixtures/groovy/grails/plugin/geb/WebDriverContainerHolder.groovy:
```diff
@@ -270,58 +367,117 @@ class WebDriverContainerHolder {
+                    return body.call()
+                } finally {
+                    if (prev == null) map.remove(key) else map[key] = prev
+                    if (map.isEmpty()) OVERRIDDEN_SYSTEM_PROPERTIES.remove()
```
Should this be removed?  What recreates the map?  Is the assumption the thread won't be reused and will always terminate?

matrei replied:
No, it should stay.
```groovy
def c = ClosureDecorators.withSystemProperty({
    println System.getProperty('my.custom.property') // should print "customValue"
}, 'my.custom.property', 'customValue')
c()
new Thread({
    sleep 1000
    c()
}).start()
println System.getProperty('my.custom.property') // should print original value, probably null
```
```console
customValue
null

customValue
```

Without it:
```console
customValue
customValue

customValue
```

jdaugherty reviewed (approved):
Great Job!  Only minor comments ...

matrei committed:
fix(geb): feedback - use `BigDecimal` in `GrailsGebSettings`

matrei committed:
ci: feedback - set `atCheckWaiting` in build file
